pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,335 @@
+    pid_t pid = ReadPidFromFile(pid_filename);
+    if (pid != -1 && IsProcessRunning(pid)) {
+      return pid;
+    }
+
+    if (i % 100 == 0) {
+      handler_->Message(
+          kWarning,
+          "Controller stuck waiting for master process PID to appear in %s",
+          pid_filename.c_str());
+    }
+
+    // Sleep instead of spinning.
+    timer_->SleepUs(10 * 1000 /* 0.01s */);
+  }
+}
+
+GoogleString ControllerManager::ControllerPidFilename() {
+  SystemRewriteOptions* options = reinterpret_cast<SystemRewriteOptions*>(
+      factory_->default_options());  // Cast is safe, since we're in system.
+  CHECK(options != NULL);
+
+  GoogleString controller_pid_filename(options->controller_pid_filename());
+  if (controller_pid_filename.empty()) {
+    handler_->Message(
+        kError, "Controller PID Filename should not be set to empty");
+    return "";
+  } else if (controller_pid_filename[0] == '/') {
+    return controller_pid_filename;
+  } else {
+    // Relative paths are interpreted relative to the directory in which the
+    // webserver would normally store the pid file for its master process.
+
+    GoogleString master_pid_filename = factory_->MasterPidFilename();
+    size_t final_slash = master_pid_filename.find_last_of('/');
+
+    return StrCat(
+        master_pid_filename.substr(0, final_slash), "/",
+        controller_pid_filename);
+  }
+}
+
+void ControllerManager::WriteControllerPid() {
+  GoogleString controller_pid_filename(ControllerPidFilename());
+  if (controller_pid_filename.empty()) {
+    return;
+  }
+
+  file_system_->WriteFile(controller_pid_filename.c_str(),
+                          Integer64ToString(getpid()),
+                          handler_);
+}
+
+pid_t ControllerManager::ControllerPid() {
+  GoogleString controller_pid_filename(ControllerPidFilename());
+  if (controller_pid_filename.empty()) {
+    return -1;
+  }
+
+  return ReadPidFromFile(controller_pid_filename);
+}
+
+void ControllerManager::KillControllerIfAlreadyRunning() {
+  pid_t controller_pid = ControllerPid();
+  if (controller_pid == -1) {
+    // Lookup failure, already logged.
+    return;
+  }
+
+  handler_->Message(
+      kInfo,
+      "About to send SIGKILL to babysitter process (%d) to tell it to "
+      "exit.  The controller will see it exit and exit as well.",
+      controller_pid);
+  // Babysitter has no cleanup to do, no reason to avoid SIGKILL.
+  int ret = kill(controller_pid, SIGKILL);
+  if (ret == -1) {
+    // Signal failed to send.
+    if (errno == ESRCH) {
+      // This is fine.  This is what usually happens if the controller exits
+      // abruptly and leaves behind a stale pid in its file.
+    } else if (errno == EPERM) {
+      handler_->Message(
+          kWarning,
+          "Permissions error trying to kill controller process at PID %d.",
+          controller_pid);
+    } else {
+      LOG(DFATAL) << "Got unexpected failure " << errno << " from kill.";
```

Comment:
+1 to not using strerror, actually, since it feels like much of the time I see the output the first step is grepping to figure out what code # it is anyway.